src/ipc/ipc4/dai.c

```diff
@@ -240,6 +240,81 @@ void dai_dma_release(struct comp_dev *dev)
 	}
 }
 
+static int dai_get_unused_llp_slot(struct comp_dev *dev,
+				   union ipc4_connector_node_id *node)
+{
+	struct ipc4_llp_reading_slot slot;
+	k_spinlock_key_t key;
+	uint32_t max_slot;
+	uint32_t offset;
+	int i;
+
+	/* sdw with multiple gateways uses sndw_reading_slots */
+	if (node->f.dma_type == ipc4_alh_link_output_class &&
+	    node->f.v_index >= ALH_MULTI_GTW_BASE) {
+		offset = SRAM_REG_LLP_SNDW_READING_SLOTS;
+		max_slot = IPC4_MAX_LLP_SNDW_READING_SLOTS - 1;
+	} else {
+		offset = SRAM_REG_LLP_GPDMA_READING_SLOTS;
+		max_slot = IPC4_MAX_LLP_GPDMA_READING_SLOTS;
+	}
+
+	key = k_spin_lock(&sof_get()->fw_reg_lock);
+
+	/* find unused llp slot offset with node_id of zero */
+	for (i = 0; i < max_slot; i++, offset += sizeof(slot)) {
+		uint32_t node_id;
+
+		node_id = mailbox_sw_reg_read(offset);
```

Comment:
where else are these slots accessed? Presumably in `dai_dma_position_update()` There's probably a location where they're freed too, right? In `dai_dma_release()`? That probably has to be protected too. And where slots are read as well at least because access to them isn't atomic - they contain two 32-bit fields, so you want reading from them to be at least consistent.

Reply:
@lyakh  We only need to protect LLP slot allocation. After allocation, the used dai can access it without problem since no other dais will access it.  The LLP slot is only read by Linux or windows driver not by FW itself and FW only save hw LLP value to it or clean it for dai_dma_release.

Reply:
I think the closest we can get to a real race condition here is during a shutdown - when streaming is running and at the same time `dai_dma_release()` is called and clears the slot with `memset_s(&slot, 0, sizeof(slot))`. So I would say that more than allocation it makes sense to lock when reading it during streaming and when releasing it during shut-down. Locking during allocation seems to be less likely to be needed, but if we want to, we can lock that too.

Reply:
@lyakh  "memset_s(&slot, 0, sizeof(slot))" here slot is local variable, not llp_slot memory.
"when reading it during streaming and when releasing it during", no llp slot reading in FW, just reading in Linux driver.   dai_dma_release is only called by dai_reset. At this time dai is paused before reset, so there is no streaming.

Reply:
yeah, sorry, I meant a couple of lines down from there:
```
mailbox_sw_regs_write(offset, &slot, sizeof(slot));
```

Reply:
@lyakh  your idea is a good view, but it never happen.
````c
void dai_dma_release(struct comp_dev *dev)
	/* cannot configure DAI while active */
	if (dev->state == COMP_STATE_ACTIVE) {
		comp_info(dev, "dai_config(): Component is in active state. Ignore resetting");
		return;
	}
....
        mailbox_sw_regs_write(llp_reg_offset, &slot, sizeof(slot));
....
````

Reply:
rrright, but this is this DAI state, but the slot array is global for all DAIs, isn't it?

Reply:
yes, it is only write the this dai, not by other dai.

Reply:
llp slot is only written by its allocation dai

Reply:
Yes, I understand that - each slot is written by its respective DAI. But the whole array is common for all DAIs? So when one DAI frees its slot, another DAI might be searching for a free slot in that array, so these two processes might race - the only thing that stops them from racing now is the IPC serialisation, right?

Reply:
If dai A is releasing slotX when dai B is checking slot, does dai B need to use the slotX free by dai A at that time ? In my opinion dai B can use any free slot

Reply:
that's why I'm saying - it's hard to discuss hypothetic problems. Currently this and many other locations are protected by IPC serialisation. Here we are deciding that that isn't enough and we want to add some explicit protection, but we don't know against what. If these two processes weren't serialised but completely asynchronous - freeing and acquiring slots, I think we could have a problem, because writing to slots isn't atomic, slots consist of 2 32-bit words. So, if DAI A is releasing slot X, it has set `.node_id` field in it to 0, at which time it's interrupted by the other process, looking for a free slot for DAI B. It finds slot X with `.node_id == 0` so it concludes that it's free, grabs it and also sets `.reg_offset`. After that the original process resumes and continues its `mailbox_sw_regs_write()` execution and overwrites `reg_offset` with 0.

Reply:
@lyakh  reg_offset is not in llp_slot, instead it is a variable of dai.  When releasing llp_slot.reading will be clear to zero, and when allocating and initialization llp_slot, the llp_slot.reading is also cleared to zero.  NP.

Reply:
so you meant that the interrupt thread will be resumed after a long time ? then it will overwrite the llp slot when llp slot is updated by dai copy in another thread ? It is very rarely but possible.